server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java

```diff
@@ -0,0 +1,168 @@
+import java.util.Map.Entry;
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import org.apache.james.jmap.exceptions.MailboxHasChildException;
+import org.apache.james.jmap.exceptions.MailboxNameException;
+import org.apache.james.jmap.exceptions.MailboxParentNotFoundException;
+import org.apache.james.jmap.model.SetError;
+import org.apache.james.jmap.model.SetMailboxesRequest;
+import org.apache.james.jmap.model.SetMailboxesResponse;
+import org.apache.james.jmap.model.SetMailboxesResponse.Builder;
+import org.apache.james.jmap.model.mailbox.Mailbox;
+import org.apache.james.jmap.model.mailbox.MailboxUpdateRequest;
+import org.apache.james.jmap.utils.MailboxUtils;
+import org.apache.james.mailbox.MailboxManager;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.exception.MailboxNotFoundException;
+import org.apache.james.mailbox.model.MailboxPath;
+import org.apache.james.mailbox.store.mail.model.MailboxId;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Splitter;
+
+public class SetMailboxesUpdateProcessor<Id extends MailboxId> implements SetMailboxesProcessor<Id> {
+
+    private final MailboxUtils<Id> mailboxUtils;
+    private final MailboxManager mailboxManager;
+
+    @Inject
+    @VisibleForTesting
+    SetMailboxesUpdateProcessor(MailboxUtils<Id> mailboxUtils, MailboxManager mailboxManager) {
+        this.mailboxUtils = mailboxUtils;
+        this.mailboxManager = mailboxManager;
+    }
+
+    @Override
+    public SetMailboxesResponse process(SetMailboxesRequest request, MailboxSession mailboxSession) {
+        SetMailboxesResponse.Builder responseBuilder = SetMailboxesResponse.builder();
+        request.getUpdate()
+            .entrySet()
+            .stream()
+            .forEach(update -> handleUpdate(update, responseBuilder, mailboxSession));
+        return responseBuilder.build();
+    }
+
+    private void handleUpdate(Entry<String, MailboxUpdateRequest> update, Builder responseBuilder, MailboxSession mailboxSession) {
+        String mailboxId = update.getKey();
+        MailboxUpdateRequest updateRequest = update.getValue();
+        try {
+            ensureMailboxName(updateRequest, mailboxSession);
+            Mailbox mailbox = getMailbox(mailboxId, updateRequest, mailboxSession);
+            ensureParent(mailbox, updateRequest, mailboxSession);
+
+            MailboxPath originMailboxPath = mailboxUtils.getMailboxPath(mailbox, mailboxSession);
+            MailboxPath destinationMailboxPath = computeNewMailboxPath(mailbox, originMailboxPath, updateRequest, mailboxSession);
+            if (!originMailboxPath.equals(destinationMailboxPath)) {
+                mailboxManager.renameMailbox(originMailboxPath, destinationMailboxPath, mailboxSession);
+            }
+            responseBuilder.updated(update.getKey());
+        } catch (MailboxNameException e) {
```

Comment:
With this pattern all responses are generated at the same place, which seems clearer to me.